supernode/daemon/mgr/peer_mgr.go

```diff
@@ -21,14 +29,160 @@ type PeerMgr interface {
-	// NOTE: update the info related for scheduler.
-	Update(ctx context.Context, peerID string, peerInfo *types.PeerInfo) error
+	List(ctx context.Context, filter map[string][]string) (peerList []*types.PeerInfo, err error)
+}
+
+// PeerManager is an implement of the interface of PeerMgr.
+type PeerManager struct {
+	peerStore *Store
+}
+
+// NewPeerManager return a new PeerManager Object.
+func NewPeerManager() (*PeerManager, error) {
+	return &PeerManager{
+		peerStore: NewStore(),
+	}, nil
+}
+
+// Register a peer and genreate a unique ID as returned.
+func (pm *PeerManager) Register(ctx context.Context, peerCreateRequest *types.PeerCreateRequest) (peerCreateResponse *types.PeerCreateResponse, err error) {
+	if peerCreateRequest == nil {
+		return nil, errors.Wrap(errorType.ErrEmptyValue, "peer create request")
+	}
+
+	ipString := peerCreateRequest.IP.String()
+	if !util.IsValidIP(ipString) {
+		return nil, errors.Wrapf(errorType.ErrInvalidValue, "peer IP: %s", ipString)
+	}
+
+	id := generatePeerID(peerCreateRequest)
+	peerInfo := &types.PeerInfo{
+		ID:       id,
+		IP:       peerCreateRequest.IP,
+		HostName: peerCreateRequest.HostName,
+		Port:     peerCreateRequest.Port,
+		Version:  peerCreateRequest.Version,
+		Created:  strfmt.DateTime(time.Now()),
+	}
+	pm.peerStore.Put(id, peerInfo)
+
+	return &types.PeerCreateResponse{
+		ID: id,
+	}, nil
+}
+
+// DeRegister a peer from p2p network.
+func (pm *PeerManager) DeRegister(ctx context.Context, peerID string) error {
+	if _, err := pm.getPeerInfo(peerID); err != nil {
+		return err
+	}
+
+	pm.peerStore.Delete(peerID)
+	return nil
+}
+
+// Get returns the peerInfo of the specified peerID.
+func (pm *PeerManager) Get(ctx context.Context, peerID string) (*types.PeerInfo, error) {
+	return pm.getPeerInfo(peerID)
+}
+
+// List returns all filtered peerInfo by filter.
+func (pm *PeerManager) List(ctx context.Context, filter map[string][]string) (
+	peerList []*types.PeerInfo, err error) {
+
+	// when filter is nil, return all values.
+	if filter == nil {
+		listResult := pm.peerStore.List()
+		peerList, err = pm.assertPeerInfoSlice(listResult)
+		if err != nil {
+			return nil, err
+		}
+		return
+	}
+
+	// validate the filter
+	err = validateFilter(filter, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// get param from filter
+	// For PeerInfo, there is no need to sort by field;
+	// and the order of insertion is used by default.
+	pageNum, _ := strconv.Atoi(filter[PAGENUM][0])
+	pageSize, _ := strconv.Atoi(filter[PAGESIZE][0])
+	sortDirect := filter[SORTDIRECT][0]
+
+	listResult := pm.peerStore.List()
+
+	var less func(i, j int) bool
+	if isASC(sortDirect) {
+		less = func(i, j int) bool {
+			peeri, ok := listResult[i].(*types.PeerInfo)
+			if !ok {
+				return false
+			}
+			peerj := listResult[j].(*types.PeerInfo)
+			if !ok {
+				return false
+			}
+			return time.Time(peeri.Created).Before(time.Time(peerj.Created))
+		}
+	} else if isDESC(sortDirect) {
+		less = func(i, j int) bool {
```

Comment:
These two functions could be combined into one, such as `func(i, j int, isASC bool)`.

Reply:
Has been updated. PATL. THX.